In [2]:
import pandas as pd

df = pd.read_csv("../data/raw/sdss_raw.csv")


print("Shape: ", df.shape)
print(df.info())
print(df.head())

Shape:  (100000, 11)
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 11 columns):
 #   Column     Non-Null Count   Dtype  
---  ------     --------------   -----  
 0   objID      100000 non-null  int64  
 1   specObjID  100000 non-null  uint64 
 2   redshift   100000 non-null  float64
 3   class      100000 non-null  object 
 4   u          100000 non-null  float64
 5   g          100000 non-null  float64
 6   r          100000 non-null  float64
 7   i          100000 non-null  float64
 8   z          100000 non-null  float64
 9   ra         100000 non-null  float64
 10  dec        100000 non-null  float64
dtypes: float64(8), int64(1), object(1), uint64(1)
memory usage: 8.4+ MB
None
                 objID             specObjID  redshift class         u  \
0  1237651251506446568   7670772665687494656  1.433770   QSO  20.78556   
1  1237655108917592122   8013191927121072128 -0.000459  STAR  17.09996   
2  1237651821098827886    6013548285

In [3]:
columns = ['u', 'g', 'r', 'i', 'z', 'redshift']
# checking for abnormalities in the dataset

df[columns].describe().round(3)

,u,g,r,i,z,redshift
count,100000.000,100000.000,100000.000,100000.000,100000.000,100000.000
mean,20.302,19.423,19.304,18.879,18.753,0.518
std,54.907,54.900,1.637,44.831,44.831,0.605
min,-9999.000,-9999.000,12.364,-9999.000,-9999.000,-0.004
25%,19.410,18.579,18.203,18.015,17.874,-0.000
50%,20.542,19.887,19.534,19.294,19.162,0.010
75%,21.669,21.023,20.607,20.353,20.192,1.083
max,31.143,31.037,24.332,30.029,27.955,2.174


In [6]:
print('amount of abnormalities in each column. my guess is they will probably be the same number.')

mask = df['u'] == -9999
print("u: ", mask.sum())

mask = df['g'] == -9999
print("g: ", mask.sum())

mask = df['i'] == -9999
print("i: ", mask.sum())

mask = df['z'] == -9999
print("z: ", mask.sum())

amount of abnormalities in each column. my guess is they will probably be the same number.
u:  3
g:  3
i:  2
z:  2


In [9]:
print(df['u'] == -9999)
print('----')
print(df[df['u'] == -9999])

0        False
1        False
2        False
3        False
4        False
         ...  
99995    False
99996    False
99997    False
99998    False
99999    False
Name: u, Length: 100000, dtype: bool
----
                     objID            specObjID  redshift class       u  \
34930  1237648705656979647  3294486009353365504 -0.000230  STAR -9999.0   
45415  1237648704589856863   340163637186422784  0.135821   QSO -9999.0   
81754  1237648704593985738  3303522620560926720 -0.000332  STAR -9999.0   

            g         r           i          z          ra       dec  
34930 -9999.0  18.57667 -9999.00000 -9999.0000  197.410924  1.052356  
45415 -9999.0  17.52662    17.00819    16.6978  212.515272  0.213964  
81754 -9999.0  18.53601    18.62557    18.6728  221.962448  0.214552  


so only row 34930 has abnormalities in all 4 columns. now i have to find the others.

In [10]:
df[df['z'] == -9999]

,objID,specObjID,redshift,class,u,g,r,i,z,ra,dec
34930,1237648705656979647,3294486009353365504,-0.000230,STAR,-9999.00000,-9999.0000,18.57667,-9999.0,-9999.0,197.410924,1.052356
95856,1237646379377885363,2970180922680305664,-0.000569,STAR,19.99464,20.5739,18.36237,-9999.0,-9999.0,55.114934,-1.263863


found all 4 rows: 34930, 45415, 81754 and 95856.

In [24]:
print(df.sort_values(['u', 'g', 'i', 'z'], ascending=False))
print('########')
bynd_25 = df[['u', 'g', 'i', 'z']] > 25
row_mask = bynd_25.any(axis=1)
print(row_mask)
print('########')
print(df[row_mask])
print('########')
print(row_mask.sum())
print('########')
print(row_mask.head())

                     objID            specObjID  redshift class           u  \
93933  1237651497367503869  4133227303399608320 -0.000155  STAR    31.14302   
29768  1237649921115488772  5785146772931497984 -0.000108  STAR    29.31007   
54032  1237654029792968729   577739242818529280 -0.000089  STAR    28.78299   
93951  1237651822719468439  5411125900128571392 -0.000236  STAR    28.62354   
82557  1237650372089020977  4249417091143454720  0.517239   QSO    28.13979   
...                    ...                  ...       ...   ...         ...   
61969  1237651753480749111   589999622739486720 -0.000147  STAR    14.62934   
64831  1237652901309579339  2308265290389022720 -0.000088  STAR    14.42220   
81754  1237648704593985738  3303522620560926720 -0.000332  STAR -9999.00000   
45415  1237648704589856863   340163637186422784  0.135821   QSO -9999.00000   
34930  1237648705656979647  3294486009353365504 -0.000230  STAR -9999.00000   

                g         r           i           z

In [ ]:
bands = ['u', 'g', 'r', 'i', 'z']
mask_faint_30 = (df[bands] > 30).any(axis=1)
df[mask_faint_30]

,objID,specObjID,redshift,class,u,g,r,i,z,ra,dec
27019,1237651249899766024,9480285435714621440,0.853148,QSO,19.85632,19.21908,15.95324,30.02886,17.79074,233.008847,53.797540
40060,1237657070628110679,8260830257734766592,-0.001116,STAR,21.72276,31.03719,17.30637,18.54431,27.95461,40.720229,-0.071389
93933,1237651497367503869,4133227303399608320,-0.000155,STAR,31.14302,19.16503,17.17342,16.37793,26.78720,116.979133,44.297107


3 rows have mag > 30. 
- 27019 'i'
- 40060 'g'
- 93933 'u'

In [34]:
neg_mag = ((df[bands] < 0) & (df[bands] != -9999)).any(axis=1)
df[neg_mag]

,objID,specObjID,redshift,class,u,g,r,i,z,ra,dec


there are 0 negative magnitudes outside of the -9999 sentinels

In [35]:
df.groupby('class')['redshift'].describe().round(4)

,count,mean,std,min,25%,50%,75%,max
class,,,,,,,,
QSO,50000.0,1.0365,0.4400,0.0191,0.6800,1.0834,1.3872,2.1738
STAR,50000.0,-0.0004,0.0003,-0.0041,-0.0006,-0.0003,-0.0002,-0.0000


In [39]:
not_quasars = ((df['class'] == 'QSO') & (df['redshift'] < 0.05))
print(not_quasars.sum())
print(df[not_quasars].head())

39
                     objID           specObjID  redshift class         u  \
4109   1237651823251423417  664408797774637056  0.035534   QSO  18.34784   
15520  1237652900229546000  513431555163580416  0.029430   QSO  17.46433   
18592  1237654880210977225  669959681657038848  0.046779   QSO  17.66185   
24930  1237652946382749732  737480418172889088  0.041995   QSO  16.94385   
26330  1237654880205209622  663185591055181824  0.035884   QSO  15.60904   

              g         r         i         z          ra        dec  
4109   16.59131  15.76277  15.30958  14.99036  227.784136   4.203957  
15520  16.23243  15.51558  15.19153  14.95788   42.303619  -8.257145  
18592  15.65703  14.72010  14.23739  13.88971  239.732497   2.809430  
24930  16.02789  15.48159  15.10477  14.84196   10.653584 -10.822789  
26330  14.61076  13.93560  13.58366  13.29420  226.485638   3.707308  


In [41]:
df[df['redshift'] > 0.1].head()

,objID,specObjID,redshift,class,u,g,r,i,z,ra,dec
0,1237651251506446568,7670772665687494656,1.433770,QSO,20.78556,20.64296,20.08354,19.78511,19.81030,221.303424,60.277896
2,1237651821098827886,601354828572026880,0.833289,QSO,19.12911,18.97812,19.03414,19.25292,19.06842,216.028829,3.068298
3,1237651250939429598,10541823497646395392,1.198927,QSO,20.59022,20.52769,20.35182,20.33298,20.56437,113.479782,34.044699
4,1237651251510771972,693587094055970816,0.396555,QSO,19.56240,19.18494,18.63504,18.38647,17.98028,235.643009,54.026770
6,1237654599953547458,564246040249657344,1.360708,QSO,19.79511,19.61674,19.45622,19.34220,19.44237,151.830448,3.098519


In [43]:
star_anomalies = ((df['class'] == 'STAR') & ((df['redshift'] < -0.005) | (df['redshift'] > 0.002)))
print('number of star anomalies: ', star_anomalies.sum())
print('duplicate objIDs: ', df['objID'].duplicated().sum())

number of star anomalies:  0
duplicate objIDs:  0


confirms `zWarning` filter works. no abnormal stars found, no duplicated objects too.

In [46]:
bad_sentinels = (df[bands] == -9999).any(axis=1)
bad_faint = (df[bands] > 30).any(axis=1)
bad_qso_z = (df['class'] == 'QSO') & (df['redshift'] < 0.05)

bad_rows = bad_faint | bad_qso_z | bad_sentinels

df_clean = df[~bad_rows].copy()

print('before: ', len(df))
print('after cleaning: ', len(df_clean))
print('diff: ', bad_rows.sum())

before:  100000
after cleaning:  99954
diff:  46


In [49]:
df_clean.to_csv("sdss_clean.csv", index=False)